e2e/aws/redshift_test.go

```diff
@@ -55,3 +65,221 @@ func testRedshiftServerless(t *testing.T) {
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP SCHEMA %q CASCADE", testSchema))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP ROLE %q", autoRole1))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP ROLE %q", autoRole2))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserKeep))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserDrop))
+	})
+	testTable := "ctf" // capture the flag :)
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE TABLE %q.%q (id int)", testSchema, testTable))
+	require.NoError(t, err)
+
+	// create the roles that Teleport will auto assign.
+	// role 1 only allows usage of the test schema.
+	// role 2 only allows select of the test table in the test schema.
+	// a user needs to have both roles to select from the test table.
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE ROLE %q", autoRole1))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE ROLE %q", autoRole2))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("GRANT USAGE ON SCHEMA %q TO ROLE %q", testSchema, autoRole1))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("GRANT SELECT ON %q.%q TO ROLE %q", testSchema, testTable, autoRole2))
+	require.NoError(t, err)
+	autoRolesQuery := fmt.Sprintf("select 1 from %q.%q", testSchema, testTable)
+
+	var pgxConnMu sync.Mutex
+	for name, test := range map[string]struct {
+		user            string
+		dbUser          string
+		query           string
+		afterConnTestFn func(t *testing.T)
+	}{
+		"iam role": {
+			user: hostUser,
+			// role/<name> is the syntax we require for redshift IAM role auth
+			dbUser: "role/" + mustGetEnv(t, redshiftIAMDBUserEnv),
+			query:  "select 1",
+		},
+		"existing user": {
+			user:   hostUser,
+			dbUser: adminUser.Name,
+			query:  "select 1",
+		},
+		"auto user keep": {
+			user:   autoUserKeep,
+			dbUser: autoUserKeep,
+			query:  autoRolesQuery,
+			afterConnTestFn: func(t *testing.T) {
+				pgxConnMu.Lock()
+				defer pgxConnMu.Unlock()
+				waitForRedshiftAutoUserDeactivate(t, ctx, conn, autoUserKeep)
+			},
+		},
+		"auto user drop": {
+			user:   autoUserDrop,
+			dbUser: autoUserDrop,
+			query:  autoRolesQuery,
+			afterConnTestFn: func(t *testing.T) {
+				pgxConnMu.Lock()
+				defer pgxConnMu.Unlock()
+				waitForRedshiftAutoUserDrop(t, ctx, conn, autoUserDrop)
+			},
+		},
+	} {
+		test := test
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Run("connect", func(t *testing.T) {
+				route := tlsca.RouteToDatabase{
+					ServiceName: db.GetName(),
+					Protocol:    defaults.ProtocolPostgres,
+					Username:    test.dbUser,
+					Database:    "dev",
+				}
+				t.Run("via proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresConnTest(t, cluster, test.user, route, test.query)
+				})
+				t.Run("via local proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresLocalProxyConnTest(t, cluster, test.user, route, test.query)
+				})
+			})
+			if test.afterConnTestFn != nil {
+				test.afterConnTestFn(t)
```

Comment:
I'm not sure this "after" is really after, the Parallel chain above is not the simplest to parse. I think you could do something like this:

```go
			t.Run("connect", func(t *testing.T) {
				t.Parallel() // Fine being parallel.

				route := tlsca.RouteToDatabase{
					ServiceName: db.GetName(),
					Protocol:    defaults.ProtocolPostgres,
					Username:    test.dbUser,
					Database:    "dev",
				}

				t.Run("via proxy", func(t *testing.T) {
					// Not parallel!
					// (...)
				})

				t.Run("via local proxy", func(t *testing.T) {
					// Not parallel!
					// (...)
				})

				// Truly after the above now.
				if test.afterConnTestFn != nil {
					test.afterConnTestFn(t)
				}
			})
```

Reply:
I wanted the subtests to run in parallel - by grouping them under "connect", the "via proxy" and "via local proxy" subtests will run in parallel with eachother but "connect" won't finish until all its subtests have finished.

the way `t.Run()` works with `t.Parallel()` in a subtest is that as soon as a test thread sees a `t.Parallel()` it will block until its parent test function returns. The parent test will then wait while all its parallel subtests run to completion.